torch/fx/experimental/symbolic_shapes.py

```diff
@@ -978,8 +1019,8 @@ def evaluate_expr(self, expr: "sympy.Expr"):
         """
         Given an expression, evaluates it, adding guards if necessary
         """
-        if len(expr.free_symbols) == 0:
-            return expr
+        # Note: do not short-circuit here without evaluating as we might need to
+        # compute FloorDiv exprs in some cases.
```

Comment:
I don't quite understand. Are you saying something like `FloorDiv(2, 3)` won't automatically simplify? Does that seem like a problem with the built-in simplifier, it should simplify past this no?

Reply:
To give a concrete example, with the above two lines reverted:

```
% pytest test/test_dynamic_shapes.py -k floordiv -v
AssertionError: Object comparison failed: 1.0 != 2.9//2.5
```

Given all the suggestions here and in the docs, we probably want to leave some things uneval'd. I'll see if I can keep this while preserving correct behavior.

Reply:
Ah, if we do not fully simplify in `evaluate_expr` (as done in this PR), we may end up in a situation where we try to do `guard_int/float` on a `FloorDiv`. That will fail because those are defined like this: `float(self.shape_env.evaluate_expr(self.expr))`. Which will simplify to `float(FloorDiv(a, b))` and fail. We even have a test for this in `TestSymNumberMagicMethods`.